File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/2552-maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) 
    {
        unordered_map<int,int>mp;
        long long ans = 0;
        long long sum = 0;

        int start = 0;
        for(int i=0;i<nums.size();i++)
        {
            mp[nums[i]]++;
            sum+=nums[i];

            if(i>=k)
            {
                mp[nums[start]]--;
                sum-=nums[start];
                
                if(mp[nums[start]] == 0)
                mp.erase(nums[start]);
                start++;
            }
            if(mp.size() == k)
                ans = max(ans,sum);
        }
        return ans;
    }
};

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/2552-maximum-sum-of-distinct-subarrays-with-length-k.py
class Solution(object):
    def maximumSubarraySum(self, nums, k):
       
        dict = {}
        n = len(nums)

        ans = 0
        currsum = 0

        for i in range(k):

            if nums[i] in dict:
                dict[nums[i]]+=1
            else:
                dict[nums[i]] = 1

            currsum += nums[i]

        if len(dict) == k:
            ans = max(ans,currsum)

        for i in range(k,n):

            currsum = currsum - nums[i-k] + nums[i]
            if dict[nums[i-k]] == 1:
                dict.pop(nums[i-k])
            else:
                dict[nums[i-k]] -=1

            if nums[i] in dict:
                dict[nums[i]]+=1
            else:
                dict[nums[i]] = 1
            
            if len(dict) == k:
                ans = max(ans,currsum)

        return ans

            